### PR TITLE
feat(date, date-range): rewrite Date component using date-fns and remove uncontrolled support FE-4459

### LINKED ISSUE #2996: Add allowedFormats to DateInput
### Current behaviour
Currently, the component seems to only support the date format that is defined by the locale we are in.
So when I'm on a DE locale, it supports `DD.MM.YYYY` or `D.M.YY`, but not `YY-MM-DD`, for example, while in an EN locale it would be reversed.

The issue here is, that users are not adhering to only typing locally recognised formatted dates. The users are typing one of the following:
```
export const DATE_FORMAT_EUROPE = 'DD.MM.YYYY';
export const DATE_FORMAT_US = 'YYYY-MM-DD';
export const DATE_FILLER_FORMATS = [
  'D.MM.YYYY',
  'DD.M.YYYY',
  'D.M.YYYY',
  'D.M.YY',
  'DD.M.YY',
  'DD.MM.YY',
  'YYYY-MM-D',
  'YYYY-M-DD',
  'YYYY-M-D',
  'YY-MM-DD',
  'YY-M-DD',
  'YY-MM-D'
];
```
This is not a completed list. For example, I can see a user going for a `MM/DD/YYYY` format, not listed above. The problem is, I have no possible way of telling the DateInput that this is the case.

Another issue is, that users are working the Numpad for quick entering of the
different numerical fields. So they also type in dates with this, in the format `DD,MM,YYYY`.
I also can't support this as the component doesn't trigger `onChange` or `onBlur` on an invalid date, which this of course is.

### Desired behaviour
The component does not only accept formats that are valid for the current locale.

### Suggested solution
One solution would be to have a Property called 'allowedFormats' or similar, that I can provide (optionally) so the DateInput doesn't convert (which it currently does, reproducible on https://carbon.sage.com/?path=/story/experimental-date-input--default) a date like `20-06-12` (YY-MM-DD) to `20.06.2012` (DD.MM.YYYY) because it expects locale dates.
Looking at the code this would be a rather big rewrite of the component, but it would be very needed for us to have a fully supported date component.

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| carbon-react     | 16.6.0
| react-scripts    | 3.3.0
| React            | 16.12.0
| Browser          | Chrome, Firefox, Safari (Mac)
| npm              |  6.12.0
| Operating System | Windows, Mac

### LINKED ISSUE #4458: Completely remove moment js
### Desired behaviour

Moment js [in their own documentation](https://momentjs.com/docs/) recommend against using the library in favour of a more modern library, except in a small range of cases. Given that dayjs is already a project dependency for cypress, can momentjs be completely removed from the project in favour of dayjs to reduce bulk and only use a single date library?

### Current behaviour

From looking at the project, it seems that moment is used only in a slim few files, and dayjs is already imported as a project dependency for cypress.

### Suggested Solution

_No response_

### CodeSandbox or Storybook URL

_No response_

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### LINKED ISSUE #4707: DateInput picker remains visible after selecting another input
### Current behaviour

When the DateInput's picker is triggered by tapping or clicking on the **calendar icon**, but a new date is not selected, you must tap (anywhere else) twice to make the picker close.

This means that if you have another input the user taps on, the date picker remains open, even though focus has shifted to another input.

The picker goes away as expected the picker is invoked through tapping on the text input part of the component.

*Note that we can trigger this bug with a tap or mouse click in our own product but only with a tap on a tablet in codesandbox - a mouse won't work*

https://user-images.githubusercontent.com/4922335/149311914-ea69e364-7386-4e65-b604-072f4fb4be30.MP4

- Sage 200 Bug 272786

### Expected behaviour

A single click outside the picker and associated input should close the picker.

Alternatively, even if two clicks were to be expected behaviour, it should still close on one click if you tap into another input.

### CodeSandbox or Storybook URL

https://codesandbox.io/s/autumn-morning-9fv70

### JIRA Ticket (Sage Only)

_No response_

### Suggested Solution

_No response_

### Carbon Version

95.1.0

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Microsoft Edge

### What Operating System are you seeing the problem on?

Windows, Android, iOS

### Anything else we should know?

- Can reproduce on desktop and tablet in our own product but only on tablet in the codesandbox
- Our original (product side) bug report was on an Android tablet, used an iPad for the carbon bug codesandbox
- On desktop in our product, (if you find a way to reproduce in codesandbox) we found that once you invoke the picker by clicking on the calendar icon, you must not mouse over the picker to be able to reproduce the bug - you must go around it instead.

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by edleeks87**
fix #2996, fix #4458, fix #4707

### Proposed behaviour

Removes `moment` package and all uses of it in carbon as it is marked as legacy by the maintainers.
Rewrites `Date` component using `date-fns/fp` utils to handle parsing dates and formatting strings.
Removes support for using component as uncontrolled: `value` and `onChange` are now required props.
Consuming projects should update their `onChange` handlers to use the `formattedValue` to update the
value state, the `rawValue` should be used to run validations and for storing in the backend: the
`rawValue` reported in the event.target.value will either a valid ISO string when the current input
value parses correctly to a date or as `null` when it cannot. Formats are no longer configurable in
the locale object passed to the `I18nProvider` and are now handled internally: all separators are
now valid input values if they can be parsed to a valid date and the the input value will be
reformatted to the current locale's preferred format `onBlur`. The `DateRange` component has been
updated to use the new implementation of `Date`.

BREAKING CHANGE: Date component is no longer class based so any previous extension of it will no
longer work. It also now only works as a controlled component and requires both `value` and
`onChange` props to work: the `onChange` handler should also use the `formattedValue` to set the
value state and `rawValue` for validation and storing in the backend. The `DateRange` can now only
be used as a controlled component, `value` and `onChange` are required and `formattedValue` should
be used to update the state.

### Current behaviour

Uses `moment` to handle parsing and formatting. Supports controlled and uncontrolled usage. Allows implementing project to pass in the formats and format via locale object. Supports a range of non-numeric date strings and only some separators.

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

All `Date` and `DateRange` stories should be tested and compared against current versions of components in master

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

Date with locales and some validation:

https://codesandbox.io/s/vibrant-stitch-xve1t?file=/src/index.js

### CODE REVIEW

**Comment by DipperTheDan** [comment]
One thing I would like to comment on is the `rawValue` and `formattedValue` if the component doesn't have a value. Just so we have a consistent approach across all our components as with `Decimal` when `allowEmptyValue` is true, `rawValue` and `formattedValue` are both empty strings when the user does not input a value. I think which ever we decide on, it should be the same across both and/or all our components. What do you think?

**Comment by edleeks87** [comment]
Updated sandbox https://codesandbox.io/s/romantic-microservice-hgrch?file=/src/index.js

File: src/components/date-range/date-range.component.js [hunk 1]
```diff
@@ -34,89 +34,55 @@ const DateRange = ({
...
-  }, [
-    getEndDate,
-    getStartDate,
-    endDateProps,
-    startDateProps,
-    localeData,
-    today,
-  ]);
-
   function usePrevious(arg) {
```
Comment (samtjo) [inline src/components/date-range/date-range.component.js:86]: Maybe not in this PR, but I think this should be made into an internal hook, as the exact same one is used in Decimal as well

File: src/components/date-range/date-range.d.ts [hunk 2]
```diff
@@ -6,20 +6,31 @@ export interface DateRangeChangeEvent {
...
       {
-        formattedValue: number;
+        formattedValue: string;
         rawValue: string;
       }
     ];
   };
 }
 
+export interface DateInputRefMap {
```
Comment (samtjo) [inline src/components/date-range/date-range.d.ts:20]: If we want to export these all from Carbon we're better off having a `export * from "./date-range";` in the index.d.ts file as well as the default export, if not we can just remove the exports here

File: src/components/date-range/date-range.d.ts [hunk 3]
```diff
@@ -74,13 +85,15 @@ export interface DateRangeProps extends MarginProps {
...
   /** An array containing the value of startDate and endDate */
-  value?: string[];
+  value: string[];
   /** When true, validation icons will be placed on labels instead of being placed on the inputs */
   validationOnLabel?: boolean;
   /** Overrides the default tooltip position */
   tooltipPosition?: "top" | "bottom" | "left" | "right";
 }
 
+declare const DateRangeContext: React.Context<DateRangeContextProps>;
```
Comment (samtjo) [inline src/components/date-range/date-range.d.ts:95]: This should be in a `date-range-context.d.ts` file as that's where `DateRangeContext` is defined

**Review by samtjo — COMMENTED** [review]
Good work @edleeks87 - this is looking much better than the old Date/DateRange components! A few comments and questions but nothing major that I can see 👍🏻

File: src/components/date/__internal__/date-picker/date-picker.d.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+export interface DatePickerProps extends Pick<DayPickerProps, "onDayClick"> {
+  /** Boolean to toggle where DatePicker is rendered in relation to the Date Input */
+  disablePortal?: boolean;
+  /** Minimum possible date YYYY-MM-DD */
+  minDate?: string;
+  /** Maximum possible date YYYY-MM-DD */
+  maxDate?: string;
+  /** Pass any props that match the DayPickerProps interface to override default behaviors */
+  pickerProps: DayPickerProps;
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date/__internal__/date-picker/date-picker.d.ts:12]: 🤦  cheers will get them updated

File: src/components/date-range/date-range.component.js (hunk 1 above)
Comment (edleeks87) [reply src/components/date-range/date-range.component.js:86]: Yeah agree, I actually have a PR for Drawer that adds as well 😄

File: src/components/date/__internal__/date-picker/date-picker.component.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  /** Pass any props that match the DayPickerProps interface to override default behaviors */
+  pickerProps: PropTypes.object,
+  /** Callback to handle mousedown event on picker */
+  pickerMouseDown: PropTypes.func,
+};
+
+/**
+ * Returns the disabled array of days specified by props maxDate and minDate
+ */
+function getDisabledDays(minDate, maxDate) {
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date/__internal__/date-picker/date-picker.component.js:133]: Yeah fair point 👍

File: src/components/date/__internal__/navbar/navbar.d.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import NavigationProps from "react-day-picker";
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date/__internal__/navbar/navbar.d.ts:1]: Ah ok I hadn't realised

File: src/components/date/date.component.js [hunk 7]
```diff
@@ -1,624 +1,367 @@
...
-  renderDatePicker = (dateRangeProps) => {
-    if (!this.state.isDatePickerOpen) return null;
+  const handleMouseDown = (ev) => {
+    if (disabled || readOnly) {
+      return;
+    }
 
-    const { visibleValue } = this.state;
-    let isoValue = "";
+    isBlurBlocked.current = true;
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date/date.component.js:207]: I'll double check as I've been doing quite a bit of context switching the last few days, but I believe it fires the blur when you click on the input icon etc so this prevents this. If I'm wrong I'll get it removed

[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date/date.component.js:207]: Yeah without it the clicking on the input icon triggers a blur event if the input has focus

File: src/components/date-range/date-range.component.js (hunk 1 above)
Comment (DipperTheDan) [reply src/components/date-range/date-range.component.js:86]: Yeah I agree. It's used in a couple of places now so probably warrants it's own custom hook 🪝

File: src/components/date-range/date-range.component.js [hunk 8]
```diff
@@ -319,11 +316,13 @@ DateRange.propTypes = {
   /** Props for the child start Date component */
   startDateProps: PropTypes.shape({
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/date-range/date-range.component.js:317]: Good idea I'll get them updated 👍

File: src/components/date/__internal__/navbar/navbar.d.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+export interface NavbarProps {
+  month?: Date;
```
Comment (samtjo) [inline src/components/date/__internal__/navbar/navbar.d.ts:2]: These top 3 props aren't in the js propTypes, I'm not sure we need them if they're just getting passed into `StyledNavbar`

File: src/components/date-range/date-range.stories.mdx [hunk 10]
```diff
@@ -183,41 +220,87 @@ It is possible to use the `tooltipPosition` to override the default placement of
...
-      }}
-      endDateProps={{
-        allowEmptyValue: true,
-      }}
-    />
+  <Story name="locale override - example implementation" parameters={{ chromatic: { disable: true } }}>
+    {() => {
+      const [state, setState] = useState(["01/10/2016", "30/10/2016"]);
+      const handleChange = ({ target }) => {
+        setState(target.value.formattedValue);
```
[the remark this answers is not in the document]
Comment (DipperTheDan) [reply src/components/date-range/date-range.stories.mdx]: All looks good now. I'll resolve this comment 😄

File: src/components/date-range/date-range.component.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (DipperTheDan) [reply src/components/date-range/date-range.component.js:324]: Description of the prop types looks a lot better now too. I'll resolve this 😄

**Review by DipperTheDan — COMMENTED** [review]
Looks good. You've addressed my comments so just address @samtjo comment and I'll give it a ✅  👍🏻

**Review by DipperTheDan — APPROVED** [review]

**Review by samtjo — APPROVED** [review]

**Review by samtjo — APPROVED** [review]

**Comment by edleeks87** [comment]
Testing sandbox https://codesandbox.io/s/keen-bush-glqbs

File: src/components/date/__internal__/utils.js [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+function hasMatchedFormat(formatString, valueString) {
+  return (
+    formatString.length === valueString.length &&
+    isMatch(formatString, valueString)
+  );
+}
+
+export function additionalYears(formatString, value) {
+  console.log(formatString);
```
Comment (samtjo) [inline src/components/date/__internal__/utils.js:35]:
```suggestion
```

**Review by DipperTheDan — APPROVED** [review]

**Review by samtjo — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Pull request merged**